engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java

```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
         }
     }
 
-    private void compileShaders(Set<ShaderProgramFeature> features) {
-        compileShader(GL20.GL_FRAGMENT_SHADER, features);
-        compileShader(GL20.GL_VERTEX_SHADER, features);
-    }
-
     private void compileAllShaderPermutations() {
-        compileShaders(Collections.<ShaderProgramFeature>emptySet());
+        int fragShaderId = compileShader(GL20.GL_FRAGMENT_SHADER, Collections.emptySet());
+        int vertShaderId = compileShader(GL20.GL_VERTEX_SHADER, Collections.emptySet());
 
-        int counter = 1;
+        if (!compileSuccess(fragShaderId) || !compileSuccess(vertShaderId)) {
 
-        for (Set<ShaderProgramFeature> permutation : Sets.powerSet(availableFeatures)) {
-            compileShaders(permutation);
-            counter++;
+            String errorMessage = String.format("Shader '%s' failed to compile.\n\n"
+                    + "Vertex Shader Info: \n%s\n"
+                    + "Fragment Shader Info: \n%s",
+                    getURI(), getLogInfo(vertShaderId), getLogInfo(fragShaderId));
+
+            throw new RuntimeException(errorMessage);
         }
-        logger.debug("Compiled {} permutations for {}.", counter, getURI());
+
+        Set<Set<ShaderProgramFeature>> allPermutations = Sets.powerSet(availableFeatures);
+
+        for (Set<ShaderProgramFeature> permutation : allPermutations) {
+            fragShaderId = compileShader(GL20.GL_FRAGMENT_SHADER, permutation);
+            vertShaderId = compileShader(GL20.GL_VERTEX_SHADER, permutation);
+
+            if (!compileSuccess(fragShaderId) || !compileSuccess(vertShaderId)) {
+                throw new RuntimeException(String.format("Shader '%s' failed to compile for feature set '%s'",
+                        getURI(), permutation));
+            }
+        }
```

Comment:
A couple of issues here:
1) can we add the log info to the error message, like it is done for the empty set above?
2) right now as soon as a shader fails an exception is thrown and the loop stops. How about continuing for all permutations and throwing an exception only at the end, if at least one shader failed to compile? In this scenario, I don't know what's more appropriate: use logger.error() to send compilation errors to the log as soon as they are generated, with the exception being passed a static string, or simply add each error to a string and pass that to exception.